### PR TITLE
[AMDGPU] New ttracedata intrinsics

### PR DESCRIPTION
**Pull request by jayfoad**
Add llvm.amdgcn.s.ttracedata and llvm.amdgcn.s.ttracedata.imm which map
directly to the corresponding instructions s_ttracedata and
s_ttracedata_imm. These are inherently whole-wave operations so any
non-uniform inputs are readfirstlaned.

### CODE REVIEW

File: llvm/include/llvm/IR/IntrinsicsAMDGPU.td [hunk 1]
```diff
@@ -1697,6 +1697,13 @@ def int_amdgcn_s_setprio :
   DefaultAttrsIntrinsic<[], [llvm_i16_ty], [ImmArg<ArgIndex<0>>, IntrNoMem,
                                 IntrHasSideEffects]>;
 
+def int_amdgcn_s_ttracedata :
+  DefaultAttrsIntrinsic<[], [llvm_i32_ty],
+                        [IntrNoMem, IntrHasSideEffects]>;
+def int_amdgcn_s_ttracedata_imm :
```
Comment (arsenm) [inline llvm/include/llvm/IR/IntrinsicsAMDGPU.td:1703]: Do we really need the imm form, or can the backend just fold the register version with appropriate constant inputs to the imm instruction?

Comment (jayfoad) [reply llvm/include/llvm/IR/IntrinsicsAMDGPU.td:1703]: We really need it. They do different things. s_ttracedata emits a 32-bit "token" into ttracedata. s_ttracedata_imm emits an 8-bit "token" (the low 8 bits of the imm16 operand).

**Comment by jayfoad** [comment]
Ping!

File: llvm/include/llvm/IR/IntrinsicsAMDGPU.td (hunk 1 above)
Comment (arsenm) [reply llvm/include/llvm/IR/IntrinsicsAMDGPU.td:1703]: This is in no way obvious from the manual descriptions

**Review by arsenm — APPROVED** [review]

File: llvm/include/llvm/IR/IntrinsicsAMDGPU.td (hunk 1 above)
Comment (jayfoad) [reply llvm/include/llvm/IR/IntrinsicsAMDGPU.td:1703]: Yeah, I think you would have to read the instruction descriptions in conjunction with the separate threadtrace documentation, and most or all of that is not public anyway.

**Pull request merged**